crates/ruff_python_parser/src/parser/expression.rs

```diff
@@ -1951,17 +1967,17 @@ impl<'src> Parser<'src> {
         }
 
         self.expect(TokenKind::In);
-        let iter = self.parse_simple_expression();
-        self.validate_comprehension_iter_and_if_expr(&iter);
+        let iter =
+            self.parse_simple_expression(AllowLambdaExpression::No, AllowStarredExpression::No);
```

Comment:
Nit: It seems that `starred` expressions are ever only allowed when lambda expressions are allowed. We could consider having a separate method for lambda/no-lambda or design the argument so that it only ever supports Lambda yes with starred yes

Reply:
Actually, I think we can remove the `AllowLambdaExpression` because it's only allowed in `parse_conditional_expression_or_higher`. So, we can have a dedicated branch for it and always throw an error if we see it at a later point. This would be similar to yield expression but without the `parse_yield_expression_or_higher` method.

```diff
diff --git a/crates/ruff_python_parser/src/parser/expression.rs b/crates/ruff_python_parser/src/parser/expression.rs
index cedc5a7da..8a1a026b5 100644
--- a/crates/ruff_python_parser/src/parser/expression.rs
+++ b/crates/ruff_python_parser/src/parser/expression.rs
@@ -208,14 +208,18 @@ impl<'src> Parser<'src> {
         &mut self,
         allow_starred_expression: AllowStarredExpression,
     ) -> ParsedExpr {
-        let start = self.node_start();
-        let parsed_expr =
-            self.parse_simple_expression(AllowLambdaExpression::Yes, allow_starred_expression);
-
-        if self.at(TokenKind::If) {
-            Expr::If(self.parse_if_expression(parsed_expr.expr, start)).into()
+        if self.at(TokenKind::Lambda) {
+            Expr::Lambda(self.parse_lambda_expr()).into()
         } else {
-            parsed_expr
+            let start = self.node_start();
+            let parsed_expr =
+                self.parse_simple_expression(AllowLambdaExpression::Yes, allow_starred_expression);
+
+            if self.at(TokenKind::If) {
+                Expr::If(self.parse_if_expression(parsed_expr.expr, start)).into()
+            } else {
+                parsed_expr
+            }
         }
     }
 
@@ -440,9 +444,7 @@ impl<'src> Parser<'src> {
             }
             TokenKind::Lambda => {
                 let lambda_expr = self.parse_lambda_expr();
-                if previous_precedence > Precedence::Initial {
-                    self.add_error(ParseErrorType::InvalidLambdaExpressionUsage, &lambda_expr);
-                }
+                self.add_error(ParseErrorType::InvalidLambdaExpressionUsage, &lambda_expr);
                 Expr::Lambda(lambda_expr).into()
             }
             TokenKind::Yield => {

```